File: apps/web/src/lib/money/calc.test.ts
import fs from "node:fs";
import path from "node:path";
import Decimal from "decimal.js";
import { describe, expect, it } from "vitest";
import { RefundExceedsTotalError, calculateCart, matchesTotal, remainingRefundable } from "./calc";

const FIXTURES_DIR = path.resolve(
  __dirname,
  "../../../../../services/pos-engine/internal/money/fixtures",
);

function loadFixture<T>(filename: string): T {
  const filePath = path.join(FIXTURES_DIR, filename);
  const raw = fs.readFileSync(filePath, "utf-8");
  return JSON.parse(raw);
}

interface CartTotalFixture {
  name: string;
  items: Array<{
    quantity: string;
    unit_price: string;
    discount: string;
  }>;
  discount: string;
  tax_rate: string;
  expect_subtotal: string;
  expect_discount_total: string;
  expect_tax_total: string;
  expect_grand_total: string;
}

interface SplitPaymentFixture {
  name: string;
  grand_total: string;
  payments: string[];
  expect_matches: boolean;
}

interface RefundFixture {
  name: string;
  grand_total: string;
  already_refunded: string;
  requested: string;
  expect_remaining: string;
  expect_error: boolean;
}

describe("Money Calculation Parity Test Suite (Go ↔ TypeScript)", () => {
  it("passes all cart totals fixture cases (pembulatan .005 & diskon 100%)", () => {
    const cases = loadFixture<CartTotalFixture[]>("cart_totals.json");
    for (const tc of cases) {
      const result = calculateCart({
        items: tc.items.map((it) => ({
          quantity: it.quantity,
          unitPrice: it.unit_price,
          discount: it.discount,
        })),
        discount: tc.discount,
        taxRate: tc.tax_rate,
      });

      expect(result.subtotal.toFixed(2)).toBe(tc.expect_subtotal);
      expect(result.discountTotal.toFixed(2)).toBe(tc.expect_discount_total);
      expect(result.taxTotal.toFixed(2)).toBe(tc.expect_tax_total);
      expect(result.grandTotal.toFixed(2)).toBe(tc.expect_grand_total);
    }
  });

  it("passes all split payments fixture cases (FR-22 split payment)", () => {
    const cases = loadFixture<SplitPaymentFixture[]>("split_payments.json");
    for (const tc of cases) {
      const isMatch = matchesTotal(tc.payments, tc.grand_total);
      expect(isMatch).toBe(tc.expect_matches);
    }
  });

  it("passes all refund fixture cases (partial refund & limit check)", () => {
    const cases = loadFixture<RefundFixture[]>("refunds.json");
    for (const tc of cases) {
      if (tc.expect_error) {
        expect(() =>
          remainingRefundable(tc.grand_total, tc.already_refunded, tc.requested),
        ).toThrow(RefundExceedsTotalError);
      } else {
        const remaining = remainingRefundable(tc.grand_total, tc.already_refunded, tc.requested);
        expect(remaining.toString()).toBe(tc.expect_remaining);
      }
    }
  });
});

File: services/pos-engine/internal/money/calc_test.go
package money

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/shopspring/decimal"
	"github.com/stretchr/testify/require"
)

// Fixture JSON dibaca, bukan di-hardcode di Go, supaya berkas yang sama bisa
// dipakai mirror TypeScript nanti saat apps/web ada (TESTING-STRATEGY.md §3A:
// "kedua hasil wajib identik, termasuk pembulatan"). Mirror TS itu sendiri
// TIDAK dikerjakan di sesi ini.

type cartTotalFixture struct {
	Name  string `json:"name"`
	Items []struct {
		Quantity  string `json:"quantity"`
		UnitPrice string `json:"unit_price"`
		Discount  string `json:"discount"`
	} `json:"items"`
	Discount            string `json:"discount"`
	TaxRate             string `json:"tax_rate"`
	ExpectSubtotal      string `json:"expect_subtotal"`
	ExpectDiscountTotal string `json:"expect_discount_total"`
	ExpectTaxTotal      string `json:"expect_tax_total"`
	ExpectGrandTotal    string `json:"expect_grand_total"`
}

func loadFixtures[T any](t *testing.T, path string) []T {
	t.Helper()
	raw, err := os.ReadFile(path)
	require.NoError(t, err)
	var cases []T
	require.NoError(t, json.Unmarshal(raw, &cases))
	return cases
}

func dec(t *testing.T, s string) decimal.Decimal {
	t.Helper()
	d, err := decimal.NewFromString(s)
	require.NoError(t, err)
	return d
}

// TestCalculate menutup dua dari empat kasus tepi wajib TESTING-STRATEGY.md §4:
// pembulatan .005 dan diskon 100%.
func TestCalculate(t *testing.T) {
	for _, tc := range loadFixtures[cartTotalFixture](t, "fixtures/cart_totals.json") {
		t.Run(tc.Name, func(t *testing.T) {
			in := Input{
				Discount: dec(t, tc.Discount),
				TaxRate:  dec(t, tc.TaxRate),
			}
			for _, it := range tc.Items {
				in.Items = append(in.Items, Item{
					Quantity:  dec(t, it.Quantity),
					UnitPrice: dec(t, it.UnitPrice),
					Discount:  dec(t, it.Discount),
				})
			}

			got := Calculate(in)

			require.True(t, dec(t, tc.ExpectSubtotal).Equal(got.Subtotal), "subtotal: want %s got %s", tc.ExpectSubtotal, got.Subtotal)
			require.True(t, dec(t, tc.ExpectDiscountTotal).Equal(got.DiscountTotal), "discount_total: want %s got %s", tc.ExpectDiscountTotal, got.DiscountTotal)
			require.True(t, dec(t, tc.ExpectTaxTotal).Equal(got.TaxTotal), "tax_total: want %s got %s", tc.ExpectTaxTotal, got.TaxTotal)
			require.True(t, dec(t, tc.ExpectGrandTotal).Equal(got.GrandTotal), "grand_total: want %s got %s", tc.ExpectGrandTotal, got.GrandTotal)
		})
	}
}

type splitPaymentFixture struct {
	Name          string   `json:"name"`
	GrandTotal    string   `json:"grand_total"`
	Payments      []string `json:"payments"`
	ExpectMatches bool     `json:"expect_matches"`
}

// TestSumPayments menutup kasus tepi wajib "pembayaran gabungan" (FR-22 split
// payment).
func TestSumPayments(t *testing.T) {
	for _, tc := range loadFixtures[splitPaymentFixture](t, "fixtures/split_payments.json") {
		t.Run(tc.Name, func(t *testing.T) {
			var payments []decimal.Decimal
			for _, p := range tc.Payments {
				payments = append(payments, dec(t, p))
			}
			got := MatchesTotal(payments, dec(t, tc.GrandTotal))
			require.Equal(t, tc.ExpectMatches, got)
		})
	}
}

type refundFixture struct {
	Name            string `json:"name"`
	GrandTotal      string `json:"grand_total"`
	AlreadyRefunded string `json:"already_refunded"`
	Requested       string `json:"requested"`
	ExpectRemaining string `json:"expect_remaining"`
	ExpectError     bool   `json:"expect_error"`
}

// TestRemainingRefundable menutup kasus tepi wajib "refund parsial".
func TestRemainingRefundable(t *testing.T) {
	for _, tc := range loadFixtures[refundFixture](t, "fixtures/refunds.json") {
		t.Run(tc.Name, func(t *testing.T) {
			remaining, err := RemainingRefundable(dec(t, tc.GrandTotal), dec(t, tc.AlreadyRefunded), dec(t, tc.Requested))
			if tc.ExpectError {
				require.ErrorIs(t, err, ErrRefundExceedsTotal)
				return
			}
			require.NoError(t, err)
			require.True(t, dec(t, tc.ExpectRemaining).Equal(remaining), "remaining: want %s got %s", tc.ExpectRemaining, remaining)
		})
	}
}
